### PR TITLE
chore: cleanup CLI pipeline [HEAD-60]

### PR DESCRIPTION
**Pull request by bastiandoetsch**
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/CONTRIBUTING.md) rules

#### What does this PR do?
This PR refactors the CLI build pipeline. It contains a couple of general cleanup tasks and unification tasks that are required after the full release of the Extensible CLI. Additional aspects to highlight are:

* Native Golang Builds on all available platforms
* Discontinue deployment of the following artifacts
  * docker-mac-signed-bundle.tar.gz
  * snyk-for-docker-desktop-darwin-arm64.tar.gz  
  * snyk-for-docker-desktop-darwin-x64.tar.gz

### CODE REVIEW

File: cliv2/scripts/shasum.ps1 [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    if ($expectedHash -eq $actualHash) {
+        Write-Output "$fileName : OK"
+    } else {
+        Write-Output "$fileName : FAILED"
+    }
+} else {
+    $hash = Get-Checksum -FilePath $FilePath -Algorithm $Algorithm
+    $hash = $hash.toLower()
+    Write-Output "$hash  $FilePath"
+}
```
Comment (bastiandoetsch) [inline cliv2/scripts/shasum.ps1:54]: Don't forget to delete this file.

File: scripts/install-snyk.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
```
Comment (bastiandoetsch) [inline scripts/install-snyk.sh:1]: This file can probably go away, too

File: .circleci/config-old.yml [hunk 3]
```diff
@@ -0,0 +1,1076 @@
+version: '2.1'
```
Comment (bastiandoetsch) [inline .circleci/config-old.yml:1]: This file can be deleted, too

File: .circleci/config.yml [hunk 4]
```diff
@@ -1,276 +1,517 @@
...
+          context: snyk-windows-signing
+          requires:
+            - build windows amd64
+          go_os: windows
+          go_arch: amd64
+          executor: win/default
+          install_deps_extension: windows-min
+      #          filters:
+      #            branches:
+      #              only: master
```
Comment (bastiandoetsch) [inline .circleci/config.yml:256]: reminder to uncomment

File: .circleci/config.yml [hunk 5]
```diff
@@ -1,276 +1,517 @@
...
+      #     command: npm run test:unit
+      # - run:
+      #     name: Running Go unit tests
+      #     working_directory: ./cliv2
+      #     command: make configure whiteboxtest
+      # - run:
+      #     name: Running Tap tests
+      #     command:
+      #       npx tap -Rspec --timeout=300 --node-arg=-r --node-arg=ts-node/register \
+      #       $(circleci tests glob "test/tap/*.test.*" | circleci tests split --split-by=timings)
```
Comment (bastiandoetsch) [inline .circleci/config.yml:419]: reminder to uncomment

File: .circleci/config.yml [hunk 6]
```diff
@@ -1,276 +1,517 @@
...
-    working_directory: /home/circleci/snyk
+  acceptance-tests-linux:
+    parameters:
+      test_snyk_command:
+        type: string
+      executor:
+        type: string
+    executor: << parameters.executor >>
     steps:
       - checkout
```
Comment (bastiandoetsch) [inline .circleci/config.yml:516]: is the checkout really needed? we have a workspace using the checked out source in previous jobs

File: .circleci/config.yml [hunk 7]
```diff
@@ -279,139 +520,142 @@ jobs:
...
-
-    working_directory: /home/circleci/snyk
+      executor:
+        type: string
+      items:
+        type: string
+        default: go gradle python elixir composer gradle@6 maven sbt
+    executor: << parameters.executor >>
     steps:
       - checkout
```
Comment (bastiandoetsch) [inline .circleci/config.yml:543]: Same here - really needed?

File: .circleci/config.yml [hunk 8]
```diff
@@ -279,139 +520,142 @@ jobs:
...
+      test_snyk_command:
+        type: string
+    executor:
+      name: win/default
+      size: large
     steps:
-      - run:
-          name: Creating temporary directory
-          command: mkdir /mnt/ramdisk/tmp
       - checkout
```
Comment (bastiandoetsch) [inline .circleci/config.yml:574]: Same here, really needed?

File: .circleci/config.yml [hunk 9]
```diff
@@ -279,139 +520,142 @@ jobs:
...
+          name: Install ShellSpec Deps
+          command: |
+            ./test/smoke/install-shellspec.sh --yes
+            sudo ln -s ~/.local/lib/shellspec/shellspec /usr/local/bin/shellspec
       - run:
-          name: Installing dependencies
-          command: npm install
+          name: Installing test fixture dependencies
+          working_directory: ./test/fixtures/basic-npm
+          command: npm i
```
Comment (bastiandoetsch) [inline .circleci/config.yml:610]: would `npm ci` maybe be better?

File: .circleci/config.yml [hunk 10]
```diff
@@ -279,139 +520,142 @@ jobs:
...
+            sudo ln -s << parameters.test_snyk_command >> /usr/local/bin/snyk
+            snyk --version
+      - run:
+          name: Running ShellSpec tests
+          working_directory: ./test/smoke
+          command: |
+            echo "Checkout the README in test/smoke folder for more details about this step"
+            unset SNYK_API
+            unset SNYK_API_KEY
+            export BREAK_SYSTEM_PACKAGES=1
```
Comment (bastiandoetsch) [inline .circleci/config.yml:623]: This line is superfluous

File: .circleci/config.yml [hunk 11]
```diff
@@ -279,139 +520,142 @@ jobs:
...
-
+      go_arch:
+        type: string
+      executor:
+        type: string
+      install_deps_extension:
+        type: string
+    executor: << parameters.executor >>
     steps:
       - checkout
```
Comment (bastiandoetsch) [inline .circleci/config.yml:638]: Needed?

File: .circleci/config.yml [hunk 12]
```diff
@@ -279,139 +520,142 @@ jobs:
...
-            - binary-releases/<< parameters.artifact >>
-            - binary-releases/<< parameters.artifact >>.sha256
+            - binary-releases/snyk-*
 
-  prepare-release:
-    executor: docker-node
+  create_deployment_artifacts:
+    executor: docker-amd64
     steps:
       - checkout
```
Comment (bastiandoetsch) [inline .circleci/config.yml:653]: Needed?

File: .circleci/config.yml [hunk 13]
```diff
@@ -424,18 +668,18 @@ jobs:
...
       - persist_to_workspace:
           root: .
           paths:
             - binary-releases
 
   pre-release:
-    executor: docker-node
+    executor: docker-amd64
     steps:
       - checkout
```
Comment (bastiandoetsch) [inline .circleci/config.yml:684]: needed?

File: .circleci/config.yml [hunk 14]
```diff
@@ -471,7 +715,7 @@ jobs:
           when: on_fail
 
   release-s3:
-    executor: docker-node
+    executor: docker-amd64
     steps:
       - checkout
```
Comment (bastiandoetsch) [inline .circleci/config.yml:720]: needed?

File: .circleci/config.yml [hunk 15]
```diff
@@ -487,7 +731,7 @@ jobs:
           when: on_fail
 
   release-github:
-    executor: docker-node
+    executor: docker-amd64
     steps:
       - checkout
```
Comment (bastiandoetsch) [inline .circleci/config.yml:736]: needed?

File: .circleci/config.yml [hunk 16]
```diff
@@ -504,7 +748,7 @@ jobs:
           when: on_fail
 
   release-npm:
-    executor: docker-node
+    executor: docker-amd64
     steps:
       - checkout
```
Comment (bastiandoetsch) [inline .circleci/config.yml:753]: needed?

File: Makefile [hunk 17]
```diff
@@ -5,11 +5,17 @@
 # Documentation: https://www.gnu.org/software/make/manual/make.html
 #
 
+WORKING_DIR = $(CURDIR)
```
Comment (bastiandoetsch) [inline Makefile:8]: $PWD would work too :)

File: Makefile [hunk 18]
```diff
@@ -5,11 +5,17 @@
...
 PKG := npx pkg ./ --compress Brotli
 BINARY_WRAPPER_DIR = ts-binary-wrapper
 EXTENSIBLE_CLI_DIR = cliv2
 BINARY_RELEASES_FOLDER_TS_CLI = binary-releases
 BINARY_OUTPUT_FOLDER = binary-releases
+SHASUM_CMD = shasum
+
+ifeq ($(GOOS), windows)
+	SHASUM_CMD = powershell $(WORKING_DIR)/cliv2/scripts/shasum.ps1
+endif
```
Comment (bastiandoetsch) [inline Makefile:18]: Not needed anymore

File: Makefile [hunk 19]
```diff
@@ -44,7 +50,7 @@ endif
 prepack: $(BINARY_OUTPUT_FOLDER)/version
 	@echo "'make prepack' was run. Run 'make clean-prepack' to rollback your package.json changes and this file." > prepack
 	npm version "$(shell cat $(BINARY_RELEASES_FOLDER_TS_CLI)/version)" --no-git-tag-version --workspaces --include-workspace-root
-	cd $(BINARY_WRAPPER_DIR) && npm version "$(shell cat $(CURDIR)/$(BINARY_RELEASES_FOLDER_TS_CLI)/version)" --no-git-tag-version --include-workspace-root
+	cd $(BINARY_WRAPPER_DIR); npm version "$(shell cat $(CURDIR)/$(BINARY_RELEASES_FOLDER_TS_CLI)/version)" --no-git-tag-version --include-workspace-root
```
Comment (bastiandoetsch) [inline Makefile:53]: I prefer using pushd/popd to have a directory stack available

File: Makefile [hunk 20]
```diff
@@ -72,20 +78,18 @@ $(BINARY_OUTPUT_FOLDER)/RELEASE_NOTES.md: prepack | $(BINARY_RELEASES_FOLDER_TS_
 
 # Generates a shasum of a target with the same name.
 # See "Automatic Variables" in GNU Make docs (linked at the top)
-%.sha256: %
-	cd $(@D); shasum -a 256 $(<F) > $(@F); shasum -a 256 -c $(@F)
+%.sha256:
+	cd $(@D); $(SHASUM_CMD) -a 256 $(<F) > $(@F); $(SHASUM_CMD) -a 256 -c $(@F)
```
Comment (bastiandoetsch) [inline Makefile:82]: needs to be changed, if above superfluous code is removed

File: Makefile [hunk 21]
```diff
@@ -150,7 +154,7 @@ $(BINARY_WRAPPER_DIR)/src/generated/sha256sums.txt:
 .PHONY: build-binary-wrapper
 build-binary-wrapper: pre-build-binary-wrapper $(BINARY_WRAPPER_DIR)/src/generated/version $(BINARY_WRAPPER_DIR)/src/generated/sha256sums.txt
 	@echo "-- Building Typescript Binary Wrapper ($(BINARY_WRAPPER_DIR)/dist/)"
-	@cd $(BINARY_WRAPPER_DIR) && npm run build
+	@cd $(BINARY_WRAPPER_DIR); npm run build
```
Comment (bastiandoetsch) [inline Makefile:157]: why semicolon?

File: Makefile [hunk 22]
```diff
@@ -186,11 +190,15 @@ pre-build: pre-build-binary-wrapper $(BINARY_RELEASES_FOLDER_TS_CLI)
 
 .PHONY: build
 build: pre-build
-	@cd $(EXTENSIBLE_CLI_DIR) && $(MAKE) build-full install bindir=$(CURDIR)/$(BINARY_OUTPUT_FOLDER) USE_LEGACY_EXECUTABLE_NAME=1
+	@cd $(EXTENSIBLE_CLI_DIR); $(MAKE) build-full install bindir=$(CURDIR)/$(BINARY_OUTPUT_FOLDER) USE_LEGACY_EXECUTABLE_NAME=1
```
Comment (bastiandoetsch) [inline Makefile:193]: Why not use $WORKING_DIR when defining it at the beginning of the makefile?

File: Makefile [hunk 23]
```diff
@@ -186,11 +190,15 @@ pre-build: pre-build-binary-wrapper $(BINARY_RELEASES_FOLDER_TS_CLI)
...
+
+.PHONY: sign
+sign:
+	@cd $(EXTENSIBLE_CLI_DIR); $(MAKE) sign BUILD_DIR=$(CURDIR)/$(BINARY_OUTPUT_FOLDER) USE_LEGACY_EXECUTABLE_NAME=1
```
Comment (bastiandoetsch) [inline Makefile:197]: CurDir?

File: cliv2/Makefile [hunk 24]
```diff
@@ -84,15 +85,21 @@ V1_DOWNLOAD_LINK = https://static.snyk.io/cli/v$(CLI_V1_VERSION_TAG)/$(V1_EXECUT
...
 SIGN_SCRIPT = sign_$(_GO_OS).sh
 ISSIGNED_SCRIPT = issigned_$(_GO_OS).sh
 EMBEDDED_DATA_DIR = $(WORKING_DIR)/internal/embedded/_data
+SPECIAL_SHELL =
 
+ifeq ($(_GO_OS), windows)
+	SPECIAL_SHELL = powershell
+	SHASUM_CMD = $(SPECIAL_SHELL) $(WORKING_DIR)/scripts/shasum.ps1
+	SIGN_SCRIPT = sign_$(_GO_OS).ps1
+endif
```
Comment (bastiandoetsch) [inline cliv2/Makefile:102]: looks redundant to the main makefile

File: Makefile (hunk 18 above)
Comment (PeterSchafer) [reply Makefile:18]: it is still used to validate the shasum of the legacy cli binary

File: cliv2/Makefile (hunk 24 above)
Comment (PeterSchafer) [reply cliv2/Makefile:102]: maybe at some point we can merge the two make files. or include a common set of code. for now it feels pragmatic enough to have it here as well.

File: cliv2/scripts/shasum.ps1 (hunk 1 above)
Comment (PeterSchafer) [reply cliv2/scripts/shasum.ps1:54]: We still need it during the build of the legacy cli, but not for the final binaries anymore.

File: Makefile (hunk 21 above)
Comment (PeterSchafer) [reply Makefile:162]: && was causing issues in powershell. Alternatively we can replace this depending on the host os.

File: .circleci/config.yml [hunk 25]
```diff
@@ -1,276 +1,540 @@
...
     resource_class: macos.x86.medium.gen2
+  macos-arm64:
+    macos:
+      # https://circleci.com/docs/2.0/testing-ios/#supported-xcode-versions
+      xcode: '14.3.1'
+    resource_class: macos.m1.large.gen1
 
 commands:
-  install_sdks_windows:
+  install-deps-windows-full:
```
Comment (j-luong) [inline .circleci/config.yml:42]: question: it looks like each command is a subset of `install-deps-windows-full`, so shouldn't caching mean we don't need multiple versions of this command?

File: .circleci/config.yml [hunk 26]
```diff
@@ -1,276 +1,540 @@
...
+
+      - regression-tests:
+          name: regression-tests
+          context: nodejs-install
+          filters:
+            branches:
+              ignore: master
+          requires:
+            - build linux amd64
+          test_snyk_command: /mnt/ramdisk/snyk/binary-releases/snyk-linux
```
Comment (j-luong) [inline .circleci/config.yml:221]: nitpick: can we swap the order of these so acceptance tests are grouped?

File: .circleci/config.yml (hunk 25 above)
Comment (PeterSchafer) [reply .circleci/config.yml:42]: Caching only relates to the downloaded installers, they still need to be executed to install the applications. The different versions of the command install different subsets of the tools for efficiency reasons.

File: .circleci/config.yml [hunk 27]
```diff
@@ -1,276 +1,540 @@
...
+      - run:
+          name: Running Tap tests
+          command:
+            npx tap -Rspec --timeout=300 --node-arg=-r --node-arg=ts-node/register \
+            $(circleci tests glob "test/tap/*.test.*" | circleci tests split --split-by=timings)
 
-    working_directory: /Users/distiller/snyk
-    environment:
-      HOMEBREW_NO_AUTO_UPDATE: 1
+  build-special-artifact:
```
Comment (j-luong) [inline .circleci/config.yml:444]: nitpick: can we rename to the plural form  `build-special-artifacts`

File: .circleci/config.yml [hunk 28]
```diff
@@ -279,139 +543,142 @@ jobs:
           name: Configuring artifact
           command: << parameters.test_snyk_command >> config set "api=${SNYK_API_KEY}"
```
Comment (j-luong) [inline .circleci/config.yml:544]: suggestion: the `SNYK_API_KEY` used here will pollute our `platform_hammerhead` org when `snyk monitor` acceptance tests are ran.

we should use the dedicated testing org here (and for all test jobs really) to avoid this. The test org should be available in the in `team_hammerhead-cli` context.

File: .circleci/config.yml [hunk 29]
```diff
@@ -279,139 +543,142 @@ jobs:
...
+          name: Installing sdks and tools via homebrew
+          command: |
+            brew install << parameters.items >>
+      - save_cache:
+          key: acceptance-tests-macos-<< parameters.items >>
+          paths:
+            - ~/Library/Caches/Homebrew
       - run:
           name: Configuring artifact
           command: << parameters.test_snyk_command >> config set "api=${SNYK_API_KEY}"
```
Comment (j-luong) [inline .circleci/config.yml:581]: suggestion: see [comment](https://github.com/snyk/cli/pull/4738/files#r1276080224)

File: .circleci/config.yml [hunk 30]
```diff
@@ -279,139 +543,142 @@ jobs:
...
-      - run:
-          name: Creating temporary directory
-          command: mkdir /mnt/ramdisk/tmp
       - checkout
       - attach_workspace:
           at: .
+      - install-deps-windows-full
       - run:
-          name: Configuring Snyk CLI
-          command: node ./bin/snyk config set "api=${SNYK_API_KEY}"
```
Comment (j-luong) [inline .circleci/config.yml:330]: suggestion: see [comment](https://github.com/snyk/cli/pull/4738/files#r1276080224)

File: .circleci/config-old.yml
Comment (j-luong) [inline .circleci/config-old.yml:1]: chore: remember to delete

File: .circleci/config.yml (hunk 6 above)
Comment (PeterSchafer) [reply .circleci/config.yml:539]: In order to reduce the workspace size, the workspace contains only the artifacts created in each job. This requires us to checkout first and restore the workspace later.

Comment (PeterSchafer) [reply .circleci/config.yml:539]: We could create a command that does this instead of duplicating these steps always.

File: .circleci/config.yml (hunk 9 above)
Comment (PeterSchafer) [reply .circleci/config.yml:633]: possible, but I assume in this case it is equivalent.

**Review by PeterSchafer — APPROVED** [review]

**Review by j-luong — APPROVED** [review]

File: cliv2/scripts/sign_windows.ps1 [hunk 31]
```diff
@@ -0,0 +1,54 @@
+param (
```
Comment (Avishagp) [inline cliv2/scripts/sign_windows.ps1:1]: What's the reason for this change from `.sha` to `.ps1`?

File: cliv2/scripts/sign_darwin.sh [hunk 32]
```diff
@@ -47,6 +47,8 @@ security import "$APPLE_SIGNING_SECRETS" -P "$APPLE_SIGNING_SECRETS_PASSWORD" -k
 rm $APPLE_SIGNING_SECRETS
 security set-key-partition-list -S apple-tool:,apple: -s -k "$KEYCHAIN_PASSWORD" "$KEYCHAIN_NAME"
 
+sleep 5
```
Comment (Avishagp) [inline cliv2/scripts/sign_darwin.sh:50]: Why did we add this step?

File: cliv2/Makefile [hunk 33]
```diff
@@ -197,31 +200,26 @@ lint:
...
 
 .PHONY: test-signature
-test-signature: $(ISSIGNED_SCRIPT)
+test-signature: ISSIGNED_SCRIPT
 
 # Typescript CLI targets
 $(V1_WORKING_DIR)/$(V1_BINARY_FOLDER)/$(V1_EXECUTABLE_NAME):
 	@echo "$(LOG_PREFIX) Building legacy CLI"
-	@cd $(V1_WORKING_DIR) && npm ci && npm run $(V1_BUILD_TYPE)
+	@cd $(V1_WORKING_DIR) && npm i && npm run $(V1_BUILD_TYPE)
```
Comment (Avishagp) [inline cliv2/Makefile:222]: Why not `ci`?

File: cliv2/scripts/sign_windows.ps1 (hunk 31 above)
Comment (PeterSchafer) [reply cliv2/scripts/sign_windows.ps1:1]: because it runs on windows powershell now. it was easier to write a powershell script instead of trying to make a shell script work on windows.

File: cliv2/Makefile (hunk 33 above)
Comment (PeterSchafer) [reply cliv2/Makefile:222]: it is an optimization to not always use clean-install. The pipeline itself does `npm ci` at the beginning, in this stage, we just want to install what is missing if anything.

File: cliv2/scripts/sign_darwin.sh (hunk 32 above)
Comment (PeterSchafer) [reply cliv2/scripts/sign_darwin.sh:50]: Flakiness of the script. It seems that some processes triggered run asynchronously, sleeping solved the flakiness immediately. But a good point to add a comment. @j-luong can you please do that?

File: .circleci/config.yml [hunk 34]
```diff
@@ -1,546 +1,505 @@
...
 executors:
   alpine:
     docker:
       - image: alpine:3.17
-  docker-node:
+  docker-amd64:
     docker:
-      - image: bastiandoetsch209/cli-build:20230613-040119
-    # Using RAM Disk. https://circleci.com/docs/2.0/executor-types/#ram-disks
+      - image: bastiandoetsch209/cli-build:20230725-114628
```
Comment (Avishagp) [inline .circleci/config.yml:17]: Is it ok to use this image? especially name wise :)
